src/router/index.ts

```diff
@@ -12,12 +13,37 @@ const router = createRouter({
     {
       path: '/about',
       name: 'about',
+      meta: { public: true },
       // route level code-splitting
       // this generates a separate chunk (About.[hash].js) for this route
       // which is lazy-loaded when the route is visited.
       component: () => import('../views/AboutView.vue')
+    },
+    {
+      path: '/signin',
+      name: 'signin',
+      meta: { public: true },
+      component: () => import('../views/SignInView.vue')
     }
   ]
 })
 
+router.beforeEach(async (to) => {
+  // routes without `meta: { public: true }` will check for the users, others won't
+  if (!to.meta.public) {
+    const currentUser = await getCurrentUser()
+
+    if (!currentUser) {
+      return {
+        path: '/signin',
+        query: {
+          // we keep the current path in the query so we can redirect to it after login
+          // with `router.push(route.query.redirect || '/')`
+          redirect: to.fullPath
+        }
+      }
+    }
+  }
+})
```

Comment:
This controls access for all routes. The `beforeEach` function is run on every navigation. 

When navigating to a private page, we must wait for the user to resolve. If no user is present, redirect to the sign-in page. The requested page is tracked with the `redirect` search param, and will be used once the user is signed in to store the original location.